Extend exports discovery to include omitted export specification.

ExternalReality opened the pull request:
All defined functions in a module should be found by haskell-emacs when
export specification is absent from module header.

Commit:
Extend exports discovery to include omitted export specification.

All defined functions in a module should be found by haskell-emacs when
export specification is absent from module header.

Conflicts:
	HaskellEmacs.hs

ExternalReality commented on HaskellEmacs.hs:
```diff
@@ -8,17 +9,20 @@ import           Control.Arrow
 import qualified Data.Text                        as T
-import           System.IO                        (hFlush, stdout)
+import qualified Language.Haskell.Exts.Syntax     as S (Name(Ident, Symbol)) 
 
```
Please forgive the unsolicited sort of the import lines

knupfer replied:
No problem (they were already sorted (according to stylish-haskell, alphabetically and ignoring `qualified'))

knupfer replied:
By the way, you should consider adding highlighting for trailing whitespace to emacs.

knupfer replied:
Besides, I recommend flycheck with haskell-ghc as a checker, you've got redundant parentheses (as well in the last PR) and a unused binding ("loc").

ExternalReality commented on HaskellEmacs.hs:
```diff
@@ -135,17 +139,43 @@ allExports :: [Text] -> (Text, [Text])
+unQalifiedName :: QName -> Maybe Name
+unQalifiedName (Qual _ name) = Just name
+unQalifiedName (UnQual name) = Just name
+unQalifiedName _             = Nothing
```
Will only use SYB stlye generic programming if boiler plate starts to get completely out of control.

knupfer replied:
The boiler plate is no problem.  But it seems that it is broken in some cases.  For example on a clean haskell-fun directory, it won't compile (it splices "main" into the import list).  Note that you have to apply a bugfix from master first (just a oneliner, the bug was introduced due to the macro related refactoring). With various modules put in haskell-fun, it doesn't even do that and is stuck on the first stage of compiling. I'll look into that in the coming two days. Perhaps it is related with the macro based refactoring and you just rebased your commits before merging?

Anyway thanks for the pull request, it looks promising

ExternalReality commented:
All good catches, will fix.

Commit:
[issue#15] Remove excessive whitespace and redundancies.

Commit:
[#15] Remove warnings and redundancies.

ExternalReality commented:
Catches were fixed. Please advise.

knupfer commented:
Thanks for your aide! I'll review it untill the weekend.

knupfer committed:
move tests to seperate dir, avoid parsing error

haskell-src-exts dislikes parsing the module of haskell-emacs, having
the tests in the same directory as the main module made haskell-emacs
try to parse itself

knupfer committed:
get module names and qualify function names

knupfer committed:
prettify

use stylish haskell and align some operators

knupfer commented:
I've merged, with some modifications, it finally worked on my side. There is only one problem I can see. Your blog example won't work because of the language pragma.  Basically, it is now much more robust but can't handle pragmas...

knupfer commented:
I've fixed the language issue

ExternalReality commented:
Hey Knupfer, that you for handling that. Its great that it all worked out.

knupfer commented:
It's great to work with your code, I've learned a lot!

Eric notifications@github.com writes:

> Hey Knupfer, that you for handling that. Its great that it all worked
> out.
> 
> --- Reply to this email directly or view it on GitHub:
> https://github.com/knupfer/haskell-emacs/pull/19#issuecomment-131402999
